### PR TITLE
Add AWQ/SmoothQuant to benchmark module

### PR DESCRIPTION
**Pull request by namgyu-youn**
**Summary**:
Add AWQ/SmoothQuant to the benchmark module. `vllm bench` didn't work and will be fixed in a future PR.
- fix: #2815 

**Test plan**:
```
./benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh awq_int4_weight_only meta-llama/Llama-3.2-1B
./benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh smoothquant_int8 meta-llama/Llama-3.2-1B
```

**Results** on A100:
```
+----------------------+--------------+--------------+--------------+--------------+
| Recipe               |   Checkpoint |     Wikitext |   Winogrande |   Winogrande |
|                      |         (GB) |   Perplexity |          Acc |       Stderr |
+======================+==============+==============+==============+==============+
| awq_int4_weight_only |         1.06 |      13.5956 |       0.6022 |       0.0138 |
+----------------------+--------------+--------------+--------------+--------------+
| smoothquant_int8     |         1.52 |      15.6757 |       0.5762 |       0.0139 |
+----------------------+--------------+--------------+--------------+--------------+
```

### CODE REVIEW

**Comment by namgyu-youn** [comment]
@pytorchbot label "topic: for developers"

also cc @vkuzo working on this benchmark :)

File: benchmarks/quantization/create_quantized_model.py [hunk 1]
```diff
@@ -9,25 +9,131 @@
 
 import torch
 from transformers import AutoModelForCausalLM, AutoTokenizer, TorchAoConfig
+from transformers.quantizers.auto import get_hf_quantizer
 from utils import string_to_config
 
+from torchao._models._eval import TransformerEvalWrapper
```
Comment (vkuzo) [inline benchmarks/quantization/create_quantized_model.py]: we were trying to get away from needing wrappers :)

is there a way to calibrate without a wrapper like this one?

File: benchmarks/quantization/utils.py [hunk 2]
```diff
@@ -47,5 +47,16 @@ def string_to_config(s):
             use_dynamic_per_tensor_scale=True,
             use_triton_kernel=True,
         )
+    elif s == "awq_int4_weight_only":
```
Comment (vkuzo) [inline benchmarks/quantization/utils.py]: either have a new function for these more complicated objects, or ensure that existing callsites are updated.  `string_to_config` should have a simple return type that doesn't vary based on the arguments.

**Comment by vkuzo** [comment]
i think it may be simpler to have a separate script for calibration based workflows, that way the no-calibration script can stay simple

File: benchmarks/quantization/create_quantized_model.py (hunk 1 above)
Comment (namgyu-youn) [reply benchmarks/quantization/create_quantized_model.py]: Yes actually we are already using `lm-eval` directly in AWQ example

https://github.com/pytorch/ao/blob/b78bd40e020436d135acd06e20012be87075fe37/torchao/prototype/awq/example.py#L98-L121

. Maybe we can also update other workflows not to use wrapper in a future PR?

Comment (namgyu-youn) [reply benchmarks/quantization/create_quantized_model.py]: Deprecation for wrapper (`TransformerEvalWrapper`): #3617

**Comment by namgyu-youn** [comment]
Updated to separate AWQ/SmoothQuant flows entirely. It can be summarized as:
- Workflow: `quantize_and_upload.py` vs. `calibration.py`
- Configs: `string_to_config()` vs. `string_to_calibration_config`

**Comment by vkuzo** [comment]
> i think it may be simpler to have a separate script for calibration based workflows, that way the no-calibration script can stay simple

to clarify, I meant a separate bash script, to keep everything simple in the main production benchmark scripts

the calibration flows are all prototype, IMO it's cleaner to keep it separate and consider merging to main flow as they mature

**Comment by namgyu-youn** [comment]
> > i think it may be simpler to have a separate script for calibration based workflows, that way the no-calibration script can stay simple
> 
> to clarify, I meant a separate bash script, to keep everything simple in the main production benchmark scripts
> 
> the calibration flows are all prototype, IMO it's cleaner to keep it separate and consider merging to main flow as they mature

Understood, separating them entirely is much better for maintenance. Could you check the updated (separated) workflows?

File: benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+set -e
+
+# Define all available calibration-based quantization recipes
+QUANT_RECIPES_ALL=(
+  "awq_int4_weight_only"
+  "smoothquant_int8"
+)
+
+# Define recipes that are known to be broken in vllm
```
Comment (vkuzo) [inline benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh]: let's just add vllm later? not useful to add it now since it does not work yet

File: benchmarks/quantization/calibration_based/create_quantized_model.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        tasks=tasks,
+        limit=limit,
+        batch_size=1,
+    )
+    quantize_(model, config_class(base_config, step="convert"), filter_fn=filter_fn)
+    load_config = config_class(base_config, step="prepare_for_loading")
+    model.config.quantization_config = TorchAoConfig(load_config)
+
+
+def quantize_model_and_save(model, recipe, base_config, output_dir, tasks, limit):
```
Comment (vkuzo) [inline benchmarks/quantization/calibration_based/create_quantized_model.py]: nit: add types to vars, and also the name for `base_config` should be something like `base_config_cls`

File: benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+MODEL_ID="${2:-meta-llama/Llama-3.1-8B}"
+
+# Get log file as third positional argument (optional)
+LOG_FILE="${3:-benchmarks/data/calibration_based/accuracy_and_performance_log.txt}"
+
+# Get calibration tasks as fourth positional argument (optional)
+CALIB_TASKS="${4:-wikitext}"
+
+# Get calibration limit as fifth positional argument (optional)
+CALIB_LIMIT="${5:-10}"
```
Comment (vkuzo) [inline benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh]: is 10 a good default? Usually people use 100-500 samples.

Comment (namgyu-youn) [reply benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh]: Low iteration is for fast experiment, even though we lose stability. I remember 128/256/512 are used in many papers, and we can lead external users by filing docs.

File: benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh (hunk 3 above)
Comment (namgyu-youn) [reply benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh]: Actually I am planning to debug into it after this PR. How about keeping this block for faster debugging?

File: benchmarks/quantization/calibration_based/create_quantized_model.py (hunk 4 above)
Comment (namgyu-youn) [reply benchmarks/quantization/calibration_based/create_quantized_model.py]: Done

File: benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh (hunk 3 above)
Comment (vkuzo) [reply benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh]: i'd say adding commented out / broken code with the intention to fix it later is a bit of an anti-pattern, it's usually cleaner to just add it in the PR where it works.  If you add it here there is no way to test it and it's a bit confusing to people who read your code.

File: benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh (hunk 5 above)
Comment (vkuzo) [reply benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh]: right, the default should be targeted towards useful accuracy numbers, not debugging

File: benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh (hunk 3 above)
Comment (namgyu-youn) [reply benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh]: Understood, thanks for pointing it out.

File: benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh (hunk 5 above)
Comment (namgyu-youn) [reply benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh]: I see, updated to 256. Let me know if a higher one (maybe 512) is needed.

**Comment by namgyu-youn** [comment]
vLLM failure log:
- AWQ: https://gist.github.com/namgyu-youn/dff3e22320b028b28f7d533727a88bb1
- SmoothQuant: https://gist.github.com/namgyu-youn/0dca97ff669cfebfcb3af522ae10ea83

**Comment by namgyu-youn** [comment]
@vkuzo could you please look again this?

File: benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+#   ./measure_accuracy_and_performance.sh awq_int4_weight_only           # Run single recipe
+#   ./measure_accuracy_and_performance.sh all meta-llama/Llama-3.2-1B    # Custom model with all recipes
+#   ./measure_accuracy_and_performance.sh awq_int4_weight_only meta-llama/Llama-3.2-8B my_log.txt  # All custom args
+#   SKIP_MODEL_CREATE=1 ./measure_accuracy_and_performance.sh all        # Skip model creation, only run eval
+#   SKIP_LM_EVAL=1 ./measure_accuracy_and_performance.sh all             # Skip lm_eval, only create models
+
+set -e
+
+# Define all available calibration-based quantization recipes
+QUANT_RECIPES_ALL=(
```
Comment (vkuzo) [inline benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh:36]: nit: this seems to be missing a baseline (bfloat16 without quantization), without a baseline it's hard to interpret whether the results of quantization are good or not

File: benchmarks/quantization/calibration_based/utils.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+#
+# This source code is licensed under the BSD 3-Clause license found in the
+# LICENSE file in the root directory of this source tree.
+from torchao.quantization import (
+    Int4WeightOnlyConfig,
+    Int8DynamicActivationInt8WeightConfig,
+)
+
+
+def string_to_calibration_config(s):
```
Comment (vkuzo) [inline benchmarks/quantization/calibration_based/utils.py:12]: nit: just inline this to your main function, since you don't have any other users of this

File: benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh (hunk 6 above)
Comment (namgyu-youn) [reply benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh:36]: How about running "None" from upstream then? Or we can run `lm-eval` and `vllm serve` directly for original model.

File: benchmarks/quantization/calibration_based/utils.py (hunk 7 above)
Comment (namgyu-youn) [reply benchmarks/quantization/calibration_based/utils.py:12]: Isn't separating base configs better for (1) safe logic (not all subclass supports AWQ/SmoothQuant yet), (2) future extensibility (like GPTQ)? 

Although mapping pattern is quiet complex, there might be only a few code reductions with inline.

File: benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh (hunk 6 above)
Comment (vkuzo) [reply benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh:36]: right, just make it easy for the user by including that logic here? 10x easier to just do it by default vs requiring the user to jump to a different script

Comment (namgyu-youn) [reply benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh:36]: I think we can lead users to upstream, even though their might be confusion. 

It's mostly because this module should run calibration flow, instead of direct `lm-eval & vllm` run like upstream. We can build skip logic for None case, but it would make it hard to maintain.

Comment (namgyu-youn) [reply benchmarks/quantization/calibration_based/measure_accuracy_and_performance.sh:36]: @vkuzo oh sorry, after careful look, I found we can just `pass` calibration-flow. Added "None" option, sorry for my bad again.

**Review by vkuzo — APPROVED** [review]

**Comment by vkuzo** [comment]
thanks for working on this! would you mind rebasing so we can run a fresh CI

**Comment by namgyu-youn** [comment]
> thanks for working on this! would you mind rebasing so we can run a fresh CI

Sure, finished rebase; could you trigger ci?

**Comment by namgyu-youn** [comment]
CI failure seems related to network issue (timeout), maybe rerun to fix?
